### PR TITLE
"Set Material" node (WIP)

### PR DESCRIPTION
**Pull request by portnov**
initial implementation.

## Addressed problem description

See discussion in #2460.

## Solution description

Add a new node.

## Preflight checklist

Put an x letter in each brackets when you're done this item:

- [ ] Code changes complete.
- [ ] Code documentation complete.
- [ ] Documentation for users complete (or not required, if user never sees these changes).
- [ ] Manual testing done. 
- [ ] Unit-tests implemented.
- [ ] Ready for merge.

### CODE REVIEW

File: nodes/object_nodes/material_index.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+import numpy as np
+
+import bpy
+from bpy.props import StringProperty, IntProperty, CollectionProperty
+
+from sverchok.node_tree import SverchCustomTreeNode
+from sverchok.data_structure import updateNode, match_long_repeat, fullList
+
+class SvMaterialEntry(bpy.types.PropertyGroup):
+    material = StringProperty()
```
[the remark this answers is not in the document]
Comment (portnov) [reply nodes/object_nodes/material_index.py]: Well, I agree. Will try.

File: nodes/object_nodes/material_index.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+                obj.data.materials.append(material)
+            else:
+                obj.data.materials[i] = material
+
+    def set_material_indices(self, obj, faces, materials):
+        material_indices = np.empty(len(obj.data.polygons), dtype=int)
+        obj.data.polygons.foreach_get("material_index", material_indices)
+        for face_index, material_index in zip(faces, materials):
+            #self.info("[%s] = %s", face_index, material_index)
+            material_indices[face_index] = material_index
```
[the remark this answers is not in the document]
Comment (portnov) [reply nodes/object_nodes/material_index.py]: Then we have to store "initial material index" somewhere...

**Comment by Durman** [comment]
Another problem. The node can not share managing by materials of objects with other nodes or something else. In the example I expect that at least one face will be dyed in different color. 

set_material2

This situation can be fixed if the node will add materials without removing existing materials. But in this case it will be hard to predict what index will get material set by this node. 
I see two ways of solving this problem:
- Assign materials by different node with material output socket from which will be possible to get actual index of requared material. 
- Add material inside this node. But switch from calling materials by index to by names or material objects.

**Comment by Durman** [comment]
@portnov Perhaps there is no need in this feature (for 2.79) since Sverchok for 2.8 is coming?

**Comment by portnov** [comment]
Agreed, will port this PR to 2.80 branch.

**Comment by portnov** [comment]
Screenshot_20191208_202840

**Comment by portnov** [comment]
@Durman please review / test the refactored nodes :)

**Comment by Durman** [comment]
material

**Comment by Durman** [comment]
It is possible to add `material index` input socket for `assign materials list` node. If it is not linked the node will work as it now. If the socket is linked it is possible to choose which materials should be assigned to an object.

**Comment by Durman** [comment]
And probably better make something like this:


Just example of logic of code:

```Python
class SvMaterialUiList(bpy.types.UIList):
    def draw_item(self, context, layout, data, item, icon, active_data, active_propname):
        if item.material:
            layout.prop(item.material, 'name', text='', emboss=False, icon_value=layout.icon(item.material))
            layout.operator('sverchok.material_remove', text='', icon='REMOVE')
        else:
            layout.operator('sverchok.material_add', text='', icon='ADD')


class SvAssignMaterialListNode(bpy.types.Node, SverchCustomTreeNode):
    def draw_buttons(self, context, layout):

        row.operator('sverchok.material_shift_up', text='', icon='TRIA_UP')
        row.operator('sverchok.material_shift_down', text='', icon='TRIA_DOWN')
        row.template_ID(self.materials[self.selected], "material")
```
Such approach can show dynamic icon of materials and change their order in a list.

**Comment by portnov** [comment]
> It is possible to add material index input socket for assign materials list node. If it is not linked the node will work as it now. If the socket is linked it is possible to choose which materials should be assigned to an object.

What is the use case?

**Comment by portnov** [comment]
> Just example of logic of code:

In this case, I am not able to access the node being updated in SvMaterialEntry.update_material() method; thus i'm not able to trigger nodetree update properly when the material is changed.

**Comment by portnov** [comment]
Screenshot_20191209_204136

**Comment by Durman** [comment]
> What is the use case?

In case if in set of objects each object has different set of materials. But probably this input would be redundant indeed. Each object can have similar list of materials but used only those of them which are needed.

> In this case, I am not able to access the node being updated

Well then it is better without dynamic material icons.))

**Comment by Durman** [comment]
Random material for each face of each object:

**Comment by portnov** [comment]
I'm currently working on proper list matching for both cases, when "all faces" is enabled or disabled...

**Comment by Durman** [comment]
When material is removed from list it does not remove from object material slot. Is it possible to fix?

**Comment by portnov** [comment]
It's possible, I think...

**Comment by Durman** [comment]
Filter I think will be better and simpler just switch off.
filter

**Comment by portnov** [comment]
Material per object:
Screenshot_20191209_213356

Material per face:
Screenshot_20191209_213452

Random materials:
Screenshot_20191209_213620

Set material for one specific face of each object:

Screenshot_20191209_214358

**Comment by portnov** [comment]
Screenshot_20191209_215109

**Comment by portnov** [comment]
I do not see how it is possible to disable the filter completely. I can redefine
```
def draw_filter(...):
    pass
```
in this case filter widgets will not be shown, but "collapse/expand filter" arrow is still visible (but expands nothing)...

**Comment by Durman** [comment]
Is it really need to have `per object` and `per faces` modes? We can just use `split` node for getting similar result.

**Comment by Durman** [comment]
Some thoughts about in which cases `all face` mode can be used. 

If `all faces` is enabled it means that we have full set of materials and we are ready to assign them to object. In this mode everything work fine. 

If `all faces` is disabled it means that we does not ready to assign material to all faces. For example we are ready to assign material only for windows of a building (picture above in this PR), and don't ready to assign material to other faces. 
In this case is it expected that other faces will get their materials later from other `assign material` node? Or this mode has another purpose?

**Comment by portnov** [comment]
By default, all faces of object have material_index of 0. So, in some cases it will be simpler to just set materials for several specific faces and leave the default material for all other.

**Comment by portnov** [comment]
> We can just use split node for getting similar result.

Well, this is so, but I can't say this is anywhere near obvious solution :)

**Comment by Durman** [comment]
What about such approach?

**Comment by Durman** [comment]
> By default, all faces of object have material_index of 0.

In this case probably (not sure) it would be nice to have opportunity to choose material which is assigned to all faces by default. 

And I would add notes in documentation that the nodes cant be stacked after each other in node tree because last node in tree will override all impacts of previous nodes (it is not obvious).

**Comment by Durman** [comment]
Close to our solution:
https://www.youtube.com/watch?time_continue=177&v=hCBqMeKcETg&feature=emb_logo

**Pull request merged**